src/Core/src/Stream/RewindableStream.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace AsyncAws\Core\Stream;
+
+use AsyncAws\Core\Exception\InvalidArgument;
+
+/**
+ * Provides a Stream that can be read several time.
+ *
+ * @internal
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+final class RewindableStream implements Stream
+{
+    private $content;
+
+    /**
+     * @var Stream
+     */
+    private $fallback;
+
+    private function __construct(Stream $content)
+    {
+        $this->content = $content;
+    }
+
+    public static function create($content): RewindableStream
+    {
+        if ($content instanceof self) {
+            return $content;
+        }
+        if ($content instanceof Stream) {
+            return new self($content);
+        }
+
+        throw new InvalidArgument(sprintf('Expect content to be a "Stream". "%s" given.', \is_object($content) ? \get_class($content) : \gettype($content)));
+    }
+
+    public function length(): ?int
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->length();
+        }
+
+        return $this->content->length();
+    }
+
+    public function stringify(): string
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->stringify();
+        }
+
+        return \implode('', \iterator_to_array($this));
+    }
+
+    public function getIterator(): \Traversable
+    {
+        if (null !== $this->fallback) {
+            yield from $this->fallback;
+
+            return;
+        }
+
+        $resource = \tmpfile();
```

Comment:
Is it overkill to check the length first? If the length is less that 1Mb then maybe we should store it in memory?

Reply:
Good idea !!

Reply:
I gone further by always using memory, and then switched to disk later when memory buffer exceeded 1Mo.